src/server/chromium.ts

```diff
@@ -161,111 +157,90 @@ function wrapTransportWithWebSocket(transport: ConnectionTransport, logger: Inne
     sessionToData.delete(sessionId);
   }
 
-  transport.onmessage = message => {
-    if (typeof message.id === 'number' && awaitingBrowserTarget.has(message.id)) {
-      const freshSocket = awaitingBrowserTarget.get(message.id)!;
-      awaitingBrowserTarget.delete(message.id);
+  const server = new WebSocketServer(transport, logger, port, {
+    onBrowserResponse(seqNum: number, source: ws, message: ProtocolResponse) {
+      if (awaitingBrowserTarget.has(seqNum)) {
+        const freshSocket = awaitingBrowserTarget.get(seqNum)!;
+        awaitingBrowserTarget.delete(seqNum);
 
-      const sessionId = message.result.sessionId;
-      if (freshSocket.readyState !== ws.CLOSED && freshSocket.readyState !== ws.CLOSING) {
-        const { queue } = socketToBrowserSession.get(freshSocket)!;
-        for (const item of queue!) {
-          item.sessionId = sessionId;
-          transport.send(item);
+        const sessionId = message.result.sessionId;
+        if (freshSocket.readyState !== ws.CLOSED && freshSocket.readyState !== ws.CLOSING) {
+          const { queue } = socketToBrowserSession.get(freshSocket)!;
+          for (const item of queue!) {
+            item.sessionId = sessionId;
+            server.sendMessageToBrowser(item, source);
+          }
+          socketToBrowserSession.set(freshSocket, { sessionId });
+          addSession(sessionId, freshSocket);
+        } else {
+          server.sendMessageToBrowserOneWay('Target.detachFromTarget', { sessionId });
+          socketToBrowserSession.delete(freshSocket);
         }
-        socketToBrowserSession.set(freshSocket, { sessionId });
-        addSession(sessionId, freshSocket);
-      } else {
-        transport.send({
-          id: ++lastSequenceNumber,
-          method: 'Target.detachFromTarget',
-          params: { sessionId }
-        });
-        socketToBrowserSession.delete(freshSocket);
+        return;
       }
-      return;
-    }
 
-    // At this point everything we care about has sessionId.
-    if (!message.sessionId)
-      return;
+      if (message.id !== -1)
+        this.onBrowserNotification(message);
```

Comment:
How can response be a notification?

Reply:
ProtocolResponse in transport terms is a JSONRpc response, which is either command response or a notification.